metrics/ci_integrations/docs/05_builds_synchronization.md

```diff
@@ -0,0 +1,180 @@
+
+According to the noticed points, the CI Integration tool considers that the environment's date and time are correct. This means that the tool doesn't attempt to solve the possible problems related to incorrect environment's date and time. Please consider this while configuring builds syncing, or automating this process! 
+
+### Sync Builds
+
+The second stage of the synchronization algorithm is to sync new builds from the `source` to the `destination`. The purpose of this stage is to populate the `destination` database with new builds and make them available in the Metrics Web Application. The algorithm performs the following steps:
+
+1. Fetch the last build from the `destination`.
+2. If the last build is `null`:
+    1. Fetch the `--initial-sync-limit` number of builds from the `source`.
+3. If the last build is not `null`:
+    1. Fetch builds from the `source` that were performed after the last synced build.
+4. Fetch coverage data for builds if necessary.
+5. Add new builds to the `destination`.
+
+Each new build can have one of the possible statuses:
+
+- `inProgress` - means that the build is running.
+- `successful` - means that the build has finished successfully.
+- `failed` - means that the build has finished with an error.
+- `unknown` - means that the build has finished with an unknown status.
+
+_**Note**: The `unknown` status doesn't always mean that the `source` doesn't know the build's status as well. In most cases, this means that the Metrics project doesn't support the build's status and maps it to the `unknown` one._
+
+The following activity diagram describes the sync builds stage:
+
+![Sync builds stage activity diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/ci_integrations_sync_doc/metrics/ci_integrations/docs/diagrams/sync_builds_stage_activity_diagram.puml)
+
+### Making Things Work
+
+The above sections describe the sync algorithm and diagram the behavior of its parts. However, you may be wondering how these parts behave more detailed and how the sync algorithm uses integrations while running.
+
+Let's consider the following class diagram that contains the main classes and interfaces that participate in the sync process:
+
+![Sync algorithm class diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/ci_integrations_sync_doc/metrics/ci_integrations/docs/diagrams/sync_algorithm_class_diagram.puml)
+
+When a user starts the synchronization process, the tool invokes the `CiIntegration.sync` method that performs sync algorithm calling `_syncInProgressBuilds` and `_syncBuilds` methods. These methods stand for re-syncing in-progress builds and syncing builds stages respectively. 
+
+The following sequence diagrams detail the two main processes of the sync algorithm:
+
+- **Re-Syncing In-Progress Builds**:
```

Comment:
I'm trying to understand clearly how timeout is going to be calculated - not sure that I understand it very well. For example what will timeout algorithm do in such case (chronological order):

Build start time b1
Sync time while build in progress s1
Build end time b2
Timeout duration t
Timeout time t1 = b1+t
Sync time s2

b1 --- s1 --- b2 --- t1 --- s2
So if we sync at s2 and during re-sync we receive completed build (failed or successful) so it appears that timeout passed (s2 - b1 > t) but we received build that is completed before timeout duration passed (b2-b1 < t )  (we are not syncing as often) - so are we going to check b2-b1 < t and then decide what to do (set unknown or the real data)? From the other diagrams, I'm assuming that we checking for timeout first right? 

We might want to set clear expectations of what timeout means - is it when (s2 - b1) > t and still in progress? Or something else?

Reply:
As we discussed recently, and as noticed in the document, the build should be timed out if **either**: 
- source failed to load the corresponding build **AND** `dateTimeNow - build.startedAt > timeoutDuration`;
- the corresponding build from the source still in progress **AND** `dateTimeNow - build.startedAt > timeoutDuration`. 

So the `should force timeout` from the activity diagrams of the above sections stands for checking both of these cases.